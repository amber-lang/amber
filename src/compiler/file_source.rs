use serde::{Deserialize, Serialize};

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Deserialize)]
pub struct FileMeta {
    pub is_import: bool,
    pub source: FileSource
}

impl FileMeta {
    pub fn file(is_import: bool) -> Self {
        Self {
            is_import,
            source: FileSource::File
        }
    }

    pub fn stdlib(is_import: bool) -> Self {
        Self {
            is_import,
            source: FileSource::Stdlib
        }
    }

    pub fn stream(is_import: bool) -> Self {
        Self {
            is_import,
            source: FileSource::Stream
        }
    }
}

impl Default for FileMeta {
    fn default() -> Self {
        Self {
            is_import: false,
            source: FileSource::default()
        }
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Deserialize, Default)]
pub enum FileSource {
    #[default]
    File,
    Stdlib,
    Stream
}

impl FileSource {
    pub fn cache_disabled(&self) -> bool {
        match &self {
            Self::Stream => true,
            _ => false
        }
    }
}