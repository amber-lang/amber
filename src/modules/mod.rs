pub mod statement;
pub mod expression;
pub mod block;
pub mod variable;
pub mod command;
pub mod condition;
pub mod shorthand;
pub mod loops;
pub mod function;
pub mod types;
pub mod imports;
pub mod main;
pub mod builtin;
pub mod prelude;

#[macro_export]
macro_rules! handle_types {
    ($enum_name:ident, [$($item:tt),*]) => {
        fn get_modules(&self) -> Vec<$enum_name> {
            vec![
                $(
                    $enum_name::$item($item::new())
                ),*
            ]
        }

        fn parse_match(&mut self, meta: &mut ParserMetadata, module: $enum_name) -> SyntaxResult {
            match module {
                $(
                    $enum_name::$item(module) => self.get(meta, module, $enum_name::$item)
                ),*
            }
        }

        fn translate_match(&self, meta: &mut TranslateMetadata, module: &$enum_name) -> FragmentKind {
            match module {
                $(
                    $enum_name::$item(module) => module.translate(meta)
                ),*
            }
        }

        fn document_match(&self, meta: &ParserMetadata, module: &$enum_name) -> String {
            match module {
                $(
                    $enum_name::$item(module) => module.document(meta)
                ),*
            }
        }
    };
}
