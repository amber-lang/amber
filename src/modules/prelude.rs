/// This is a prelude module that re-exports all the necessary modules for syntax modules
pub use crate::docs::module::DocumentationModule;
pub use crate::translate::fragments::block::BlockFragment;
pub use crate::translate::fragments::comment::CommentFragment;
pub use crate::translate::fragments::eval::EvalFragment;
pub use crate::translate::fragments::fragment::{FragmentRenderable, FragmentKind};
pub use crate::translate::fragments::interpolable::{InterpolableFragment, InterpolableRenderType};
pub use crate::translate::fragments::list::ListFragment;
pub use crate::translate::fragments::raw::RawFragment;
pub use crate::translate::fragments::subprocess::SubprocessFragment;
pub use crate::translate::fragments::var::{VarFragment, VarRenderType};
pub use crate::translate::module::TranslateModule;
pub use crate::utils::{ParserMetadata, TranslateMetadata};
