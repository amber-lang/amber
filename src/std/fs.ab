import { match_regex, join, replace_regex, split } from "std/text"

/// Checks if a directory exists.
pub fun dir_exists(path) {
    $ [ -d "{path}" ] $ failed {
        return false
    }
    return true
}

/// Checks if a file exists.
pub fun file_exists(path) {
    $ [ -f "{path}" ] $ failed {
        return false
    }
    return true
}

/// Gets file contents from a path.
pub fun file_read(path) {
    return $ < "{path}" $?
}

/// Writes content to a file.
/// Doesn't check if the file exist
pub fun file_write(path, content) {
    return $ echo "{content}" > "{path}" $?
}

/// Appends content to a file.
///
/// Doesn't check if the file exists.
pub fun file_append(path, content) {
    return $ echo "{content}" >> "{path}" $?
}

/// Creates a symbolic link.
///
/// If the file doesn't exist, it returns a boolean and prints a message.
pub fun symlink_create(origin: Text, destination: Text): Bool {
    if file_exists(origin) {
        trust $ ln -s "{origin}" "{destination}" $
        return true
    }

    echo "The file {origin} doesn't exist!"
    return false
}

/// Creates a directory with all parent directories as required.
pub fun dir_create(path: Text): Null {
    if not dir_exists(path) {
        trust $ mkdir -p "{path}" $
    }
}

/// Changes the permission bits of a file.
///
/// If the file doesn't exist, it returns `false` and prints a message.
pub fun file_chmod(path: Text, mode: Text): Bool {
    if file_exists(path) {
        trust $ chmod "{mode}" "{path}" $
        return true
    }

    echo "The file {path} doesn't exist!"
    return false
}

/// Changes the owner of a file.
///
/// If the file doesn't exist, it returns `false` and prints a message.
pub fun file_chown(path: Text, user: Text): Bool {
    if file_exists(path) or dir_exists(path) {
        trust $ chown -R "{user}" "{path}" $
        return true
    }

    echo "The file {path} doesn't exist!"
    return false
}

/// Escapes all characters in the passed-in glob except "*", "?" and "/",
/// to prevent injection attacks.
fun escape_non_glob_chars(path: Text): Text {
    return replace_regex(path, "\([^*?/]\)", "\\\\\1")
}

/// Finds all files or directories matching multiple file globs. When
/// we have union types, this functionality can be merged into the main
/// `glob` function.
pub fun file_glob_all(paths: [Text]): [Text]? {
    let combined = ""
    if len(paths) == 1 {
        combined = escape_non_glob_chars(paths[0])
    } else {
        let items = [Text]
        for item in paths {
            item = escape_non_glob_chars(item)
            items += [item]
        }
        combined = join(items, " ")
    }
    let files = $ eval "for file in {combined}; do [ -e \\\"\\\$file\\\" ] && echo \\\"\\\$file\\\"; done" $?
    return split(files, "\n")
}

/// Finds all files or directories matching a file glob.
pub fun file_glob(path: Text): [Text]? {
    return file_glob_all([path])?
}

/// Extract the file detecting from the filename the extension
/// Supports: bz2, gz, xz, bz2, deb, rar, rpm, tar(gz/xz/bz), zip(war/jar), 7z
/// Note: Not all the commands supports the output folder path
pub fun file_extract(path: Text, target: Text): Null? {
    if file_exists(path) {
        if {
            match_regex(path, "\.\(tar\.bz2\|tbz\|tbz2\)$"): $ tar xvjf "{path}" -C "{target}" $?
            match_regex(path, "\.\(tar\.gz\|tgz\)$"): $ tar xzf "{path}" -C "{target}" $?
            match_regex(path, "\.\(tar\.xz\|txz$\)$"): $ tar xJf "{path}" -C "{target}" $?
            match_regex(path, "\.bz2$"): $ bunzip2 "{path}" $?
            match_regex(path, "\.deb$"): $ dpkg-deb -xv "{path}" "{target}" $?
            match_regex(path, "\.gz$"): $ gunzip "{path}" $?
            match_regex(path, "\.rar$"): $ unrar x "{path}" "{target}" $?
            match_regex(path, "\.rpm$"): $ rpm2cpio "{path}" | cpio -idm $?
            match_regex(path, "\.tar$"): $ tar xf "{path}" -C "{target}" $?
            match_regex(path, "\.xz$"): $ xz --decompress "{path}" $?
            match_regex(path, "\.7z$"): $ 7z -y "{path}" -o "{target}" $?
            match_regex(path, "\.\(zip\|war\|jar\)$"): $ unzip "{path}" -d "{target}" $?
            else {
                echo "Error: Unsupported file type"
                fail 3
            }
        }
    } else {
        echo "Error: File not found"
        fail 2
    }
}
