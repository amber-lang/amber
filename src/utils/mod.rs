pub mod cc_flags;
pub mod context;
pub mod function_cache;
pub mod function_interface;
pub mod function_metadata;
pub mod import_cache;
pub mod metadata;
pub mod payload;

pub use metadata::*;
