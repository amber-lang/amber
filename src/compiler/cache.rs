use std::{fs::{self, Permissions}, os::unix::fs::PermissionsExt, path::PathBuf};

const GIT_HASH: &'static str = env!("GIT_HASH");

pub mod parse;
pub mod tokenize;

pub fn home_cache() -> Option<PathBuf> {
    if let Some(mut home) = home::home_dir() {
        home.push(".cache");
        home.push("amber");
        if ! home.is_dir() {
            fs::create_dir_all(&home).expect("Couldn't create ~/.cache/amber");

            #[cfg(unix)]
            fs::set_permissions(&home, Permissions::from_mode(0o700)).expect("Couldn't set permissions to ~/.cache/amber")
        }
        Some(home)
    } else {
        None
    }
}
