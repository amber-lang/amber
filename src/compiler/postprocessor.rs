use std::cell::RefCell;
use std::io::{BufWriter, Write};
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::rc::Rc;

use itertools::Itertools;
use wildmatch::WildMatchPattern;

#[derive(Debug, Clone)]
pub struct PostProcessor {
    pub name: String,
    pub bin: PathBuf,
    command: Rc<RefCell<Command>>,
}

impl PostProcessor {
    pub fn new<N: Into<String>, B: Into<PathBuf>>(name: N, bin: B) -> Self {
        let name: String = name.into();
        let bin: PathBuf = bin.into();
        let mut command = Command::new(bin.clone());
        command.stdin(Stdio::piped());
        command.stdout(Stdio::piped());
        command.stderr(Stdio::piped());
        let command = Rc::new(RefCell::new(command));
        Self {
            name,
            bin,
            command,
        }
    }

    pub fn cmd(&self) -> Rc<RefCell<Command>> {
        self.command.clone()
    }

    pub fn is_available(&self) -> bool {
        match Command::new(self.bin.clone()).spawn() {
            Ok(mut v) => {
                v.kill().is_ok()
            },
            Err(_) => false
        }
    }

    pub fn execute(&self, code: String) -> Result<String, Box<dyn std::error::Error>> {
        if !self.is_available() { return Ok(code) }

        let mut spawned = self.cmd().borrow_mut().spawn()?;
        
        // send to stdin
        if let Some(stdin) = spawned.stdin.as_mut() {
            let mut writer = BufWriter::new(stdin);
            writer.write_all(code.as_bytes())?;
            writer.flush()?;
        } else {
            return Err(String::new().into())
        }

        // read from stdout or stderr
        let res = spawned.wait_with_output()?;
        if res.status.success() {
            let stdout = String::from_utf8(res.stdout)?;
            Ok(stdout)
        } else {
            let stderr = String::from_utf8(res.stderr)?;
            Err(stderr.into())
        }
    }

    pub fn get_default() -> Vec<Self> {
        let mut postprocessors = Vec::new();

        let shfmt = PostProcessor::new("shfmt", "shfmt");
        shfmt.cmd().borrow_mut().arg("-i").arg("4");
        shfmt.cmd().borrow_mut().arg("-ln").arg("bash");
        postprocessors.push(shfmt);

        let bshchk = PostProcessor::new("bshchk", "bshchk");
        bshchk.cmd().borrow_mut().arg("--ignore-shebang");
        postprocessors.push(bshchk);

        postprocessors
    }

    pub fn filter_default(filters: Vec<WildMatchPattern<'*', '?'>>) -> Vec<Self> {
        let default = Self::get_default();

        default
            .iter()
            .filter(|x| {
                filters.iter()  
                    .all(|xx| !xx.matches(&x.name))
            })
            .cloned()
            .collect_vec()
    }
}
