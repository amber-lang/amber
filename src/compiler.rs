extern crate chrono;
use crate::docs::module::DocumentationModule;
use crate::modules::block::Block;
use crate::translate::check_all_blocks;
use crate::translate::module::TranslateModule;
use crate::utils::{ParserMetadata, TranslateMetadata};
use crate::rules;
use postprocessor::PostProcessor;
use chrono::prelude::*;
use colored::Colorize;
use heraclitus_compiler::prelude::*;
use itertools::Itertools;
use wildmatch::WildMatchPattern;
use std::env;
use std::ffi::OsStr;
use std::fs;
use std::fs::File;
use std::io::{ErrorKind, Write};
use std::iter::once;
use std::path::{Path, PathBuf};
use std::process::{Command, ExitStatus};
use std::time::Instant;

pub mod postprocessor;

const NO_CODE_PROVIDED: &str = "No code has been provided to the compiler";
const AMBER_DEBUG_PARSER: &str = "AMBER_DEBUG_PARSER";
const AMBER_DEBUG_TIME: &str = "AMBER_DEBUG_TIME";

pub struct CompilerOptions {
    pub no_proc: Vec<String>,
    pub minify: bool,
    pub run_name: Option<String>,
}

impl Default for CompilerOptions {
    fn default() -> Self {
        let no_proc = vec![String::from("*")];
        Self { no_proc, minify: false, run_name: None }
    }
}

impl CompilerOptions {
    pub fn from_args(no_proc: &[String], minify: bool, input: Option<&Path>) -> Self {
        let no_proc = no_proc.to_owned();
        let run_name = input.and_then(Path::file_name)
            .and_then(OsStr::to_str)
            .map(String::from);
        Self { no_proc, minify, run_name }
    }
}

pub struct AmberCompiler {
    pub cc: Compiler,
    pub path: Option<String>,
    pub options: CompilerOptions,
}

impl AmberCompiler {
    pub fn new(code: String, path: Option<String>, options: CompilerOptions) -> AmberCompiler {
        let cc = Compiler::new("Amber", rules::get_rules());
        let compiler = AmberCompiler { cc, path, options };
        compiler.load_code(AmberCompiler::comment_shebang(code))
    }

    fn comment_shebang(code: String) -> String {
        if code.starts_with("#!") {
            String::from("// ") + &code
        } else {
            code
        }
    }

    fn env_flag_set(flag: &str) -> bool {
        if let Ok(value) = env::var(flag) {
            value == "1" || value == "true"
        } else {
            false
        }
    }

    pub fn load_code(mut self, code: String) -> Self {
        self.cc.load(code);
        self
    }

    pub fn tokenize(&self) -> Result<Vec<Token>, Message> {
        let time = Instant::now();
        match self.cc.tokenize() {
            Ok(tokens) => {
                if Self::env_flag_set(AMBER_DEBUG_TIME) {
                    let pathname = self.path.clone().unwrap_or(String::from("unknown"));
                    println!(
                        "[{}]\tin\t{}ms\t{pathname}",
                        "Tokenize".cyan(),
                        time.elapsed().as_millis()
                    );
                }
                Ok(tokens)
            }
            Err((err_type, pos)) => {
                let error_message = match err_type {
                    LexerErrorType::Singleline => {
                        format!("Singleline {} not closed", pos.data.as_ref().unwrap())
                    }
                    LexerErrorType::Unclosed => {
                        format!("Unclosed {}", pos.data.as_ref().unwrap())
                    }
                };
                let code = self.cc.code.as_ref().expect(NO_CODE_PROVIDED).clone();
                let meta = ParserMetadata::new(vec![], self.path.clone(), Some(code));
                Err(Message::new_err_at_position(&meta, pos).message(error_message))
            }
        }
    }

    pub fn parse(&self, tokens: Vec<Token>) -> Result<(Block, ParserMetadata), Message> {
        let code = self.cc.code.as_ref().expect(NO_CODE_PROVIDED).clone();
        let mut meta = ParserMetadata::new(tokens, self.path.clone(), Some(code));
        if let Err(Failure::Loud(err)) = check_all_blocks(&meta) {
            return Err(err);
        }
        let mut block = Block::new();
        let time = Instant::now();
        // Parse with debug or not
        let result = if Self::env_flag_set(AMBER_DEBUG_PARSER) {
            block.parse_debug(&mut meta)
        } else {
            block.parse(&mut meta)
        };
        if Self::env_flag_set(AMBER_DEBUG_TIME) {
            let pathname = self.path.clone().unwrap_or(String::from("unknown"));
            println!(
                "[{}]\tin\t{}ms\t{pathname}",
                "Parsed".blue(),
                time.elapsed().as_millis()
            );
        }
        // Return result
        match result {
            Ok(()) => Ok((block, meta)),
            Err(failure) => Err(failure.unwrap_loud()),
        }
    }

    fn get_sorted_ast_forest(
        &self,
        block: Block,
        meta: &ParserMetadata,
    ) -> Vec<(String, Block)> {
        let imports_sorted = meta.import_cache.topological_sort();
        let imports_blocks = meta.import_cache.files.iter()
            .map(|file| {
                file.metadata
                    .as_ref()
                    .map(|meta| (file.path.clone(), meta.block.clone()))
            })
            .collect::<Vec<Option<(String, Block)>>>();
        let mut result = vec![];
        for index in imports_sorted.iter() {
            if let Some((path, block)) = imports_blocks[*index].clone() {
                result.push((path, block));
            }
        }
        result.push((self.path.clone().unwrap_or(String::from("unknown")), block));
        result
    }

    pub fn translate(&self, block: Block, meta: ParserMetadata) -> Result<String, Message> {
        let ast_forest = self.get_sorted_ast_forest(block, &meta);
        let mut meta_translate = TranslateMetadata::new(meta, &self.options);
        let time = Instant::now();
        let mut result = vec![];
        for (_path, block) in ast_forest {
            result.push(block.translate(&mut meta_translate));
        }
        if Self::env_flag_set(AMBER_DEBUG_TIME) {
            let pathname = self.path.clone().unwrap_or(String::from("unknown"));
            println!(
                "[{}]\tin\t{}ms\t{pathname}",
                "Translate".magenta(),
                time.elapsed().as_millis()
            );
        }

        let mut result = result.join("\n") + "\n";

        let filters = self.options.no_proc.iter()
            .map(|x| WildMatchPattern::new(x))
            .collect();
        let postprocessors = PostProcessor::filter_default(filters);
        for postprocessor in postprocessors {
            result = match postprocessor.execute(result) {
                Ok(result) => result,
                Err(error) => {
                    let error = format!(
                        "Postprocessor '{}' failed\n{}",
                        postprocessor.name,
                        error.to_string().trim_end(),
                    );
                    return Err(Message::new_err_msg(error));
                },
            };
        }

        let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        let header = include_str!("header.sh")
            .replace("{{ version }}", env!("CARGO_PKG_VERSION"))
            .replace("{{ date }}", now.as_str());
        Ok(format!("{}{}", header, result))
    }

    pub fn document(&self, block: Block, meta: ParserMetadata, output: Option<String>) {
        let base_path = meta.get_path()
            .map(PathBuf::from)
            .expect("Input file must exist in docs generation");
        let base_dir = fs::canonicalize(base_path).map(|val| {
            val.parent()
                .expect("Parent dir must exist in docs generation")
                .to_owned()
                .clone()
        });
        let base_dir = base_dir.unwrap_or_else(|err| {
            Message::new_err_msg("Couldn't get the absolute path to the provided input file")
                .comment(err.to_string())
                .show();
            std::process::exit(1);
        });
        let ast_forest = self.get_sorted_ast_forest(block, &meta);
        let mut paths = vec![];
        for (path, block) in ast_forest {
            let dep_path = {
                let dep_path = match fs::canonicalize(PathBuf::from(path)) {
                    Ok(path) => path,
                    Err(_) => continue,
                };
                if !dep_path.starts_with(&base_dir) {
                    continue;
                }
                dep_path
            };
            let document = block.document(&meta);
            // Check if an output directory was specified.
            if let Some(output) = &output {
                // Save to file; replace the base directory if the output
                // path is absolute, otherwise append the output path.
                let dir_path = {
                    let file_path = dep_path.strip_prefix(&base_dir).unwrap();
                    let file_dir = file_path.parent().unwrap();
                    base_dir.join(output).join(file_dir)
                };
                if let Err(err) = fs::create_dir_all(dir_path.clone()) {
                    let message = format!("Couldn't create directory `{}`. Do you have sufficient permissions?", dir_path.display());
                    Message::new_err_msg(message)
                        .comment(err.to_string())
                        .show();
                    std::process::exit(1);
                }
                let filename = dep_path.file_stem().unwrap().to_string_lossy();
                let path = dir_path.join(format!("{filename}.md"));
                let mut file = File::create(path.clone()).unwrap();
                file.write_all(document.as_bytes()).unwrap();
                paths.push(String::from(path.to_string_lossy()));
            } else {
                // Write to standard output.
                std::io::stdout().write_all(document.as_bytes()).unwrap();
            }
        }
        if !paths.is_empty() {
            let files = if paths.len() > 1 { "Files" } else { "File" };
            let message = once(format!("{files} generated at:")).chain(paths).join("\n");
            Message::new_info_msg(message).show();
        }
    }

    pub fn compile(&self) -> Result<(Vec<Message>, String), Message> {
        let tokens = self.tokenize()?;
        let (block, meta) = self.parse(tokens)?;
        let messages = meta.messages.clone();
        let code = self.translate(block, meta)?;
        Ok((messages, code))
    }

    pub fn execute(mut code: String, args: Vec<String>) -> Result<ExitStatus, std::io::Error> {
        if let Some(mut command) = Self::find_bash() {
            if !args.is_empty() {
                let args = args.into_iter()
                    .map(|arg| arg.replace("\"", "\\\""))
                    .map(|arg| format!("\"{arg}\""))
                    .collect::<Vec<String>>();
                code = format!("set -- {}\n{}", args.join(" "), code);
            }
            command.arg("-c").arg(code).spawn()?.wait()
        } else {
            let error = std::io::Error::new(ErrorKind::NotFound, "Failed to find Bash");
            Err(error)
        }
    }

    pub fn generate_docs(&self, output: Option<String>, usage: bool) -> Result<(), Message> {
        let tokens = self.tokenize()?;
        let (block, mut meta) = self.parse(tokens)?;
        meta.doc_usage = usage;
        self.document(block, meta, output);
        Ok(())
    }

    #[cfg(test)]
    pub fn test_eval(&mut self) -> Result<String, Message> {
        self.options.no_proc = vec!["*".into()];
        self.compile().map_or_else(Err, |(_, code)| {
            if let Some(mut command) = Self::find_bash() {
                let child = command.arg("-c").arg::<&str>(code.as_ref()).output().unwrap();
                let output = String::from_utf8_lossy(&child.stdout).to_string();
                Ok(output)
            } else {
                let message = Message::new_err_msg("Failed to find Bash");
                Err(message)
            }
        })
    }

    #[cfg(windows)]
    fn find_bash() -> Option<Command> {
        if let Some(paths) = env::var_os("PATH") {
            for path in env::split_paths(&paths) {
                let path = path.join("bash.exe");
                if path.exists() {
                    let command = Command::new(path);
                    return Some(command);
                }
            }
        }
        return None;
    }

    #[cfg(not(windows))]
    fn find_bash() -> Option<Command> {
        let mut command = Command::new("/usr/bin/env");
        command.arg("bash");
        Some(command)
    }
}
