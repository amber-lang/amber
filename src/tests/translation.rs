use crate::compiler::file_source::FileMeta;
/// Tests for Amber scripts that check snapshot of generated AST.
use crate::compiler::{AmberCompiler, CompilerOptions};
use crate::modules::prelude::{FragmentKind, TranslateModule};
use crate::utils::TranslateMetadata;
use test_generator::test_resources;
use insta::assert_debug_snapshot;
use std::fs;
use std::path::Path;

pub fn translate_amber_code<T: Into<String>>(code: T) -> Option<FragmentKind> {
    let options = CompilerOptions::default();
    let compiler = AmberCompiler::new(code.into(), None, options, FileMeta::stream(true));
    let tokens = compiler.tokenize().ok()?;
    let (ast, meta) = compiler.parse(tokens).ok()?;
    let mut translate_meta = TranslateMetadata::new(meta, &compiler.options);
    Some(ast.translate(&mut translate_meta))
}

/// Autoload the Amber test files in translation
#[test_resources("src/tests/translation/*.ab")]
fn test_translation(input: &str) {
    let code = fs::read_to_string(input)
        .unwrap_or_else(|_| panic!("Failed to open {input} test file"));
    let ast = translate_amber_code(code)
        .expect("Couldn't translate Amber code");
    let filename = Path::new(input).file_name()
        .expect("Provided directory")
        .to_str()
        .expect("Cannot translated to string");
    assert_debug_snapshot!(filename, ast);
}
