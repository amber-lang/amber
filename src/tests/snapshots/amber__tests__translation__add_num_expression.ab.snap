---
source: src/tests/translation.rs
expression: ast
---
Block(
    BlockFragment {
        statements: [
            List(
                ListFragment {
                    values: [
                        Raw(
                            RawFragment {
                                value: "echo ",
                            },
                        ),
                        Subprocess(
                            SubprocessFragment {
                                fragment: List(
                                    ListFragment {
                                        values: [
                                            Raw(
                                                RawFragment {
                                                    value: "echo ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "15.15",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " '+' ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "45.15",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | bc ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "-l",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | sed '",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "'",
                                                },
                                            ),
                                        ],
                                        separator: Empty,
                                    },
                                ),
                                quoted: true,
                            },
                        ),
                    ],
                    separator: Empty,
                },
            ),
            List(
                ListFragment {
                    values: [
                        Raw(
                            RawFragment {
                                value: "echo ",
                            },
                        ),
                        Subprocess(
                            SubprocessFragment {
                                fragment: List(
                                    ListFragment {
                                        values: [
                                            Raw(
                                                RawFragment {
                                                    value: "echo ",
                                                },
                                            ),
                                            Subprocess(
                                                SubprocessFragment {
                                                    fragment: List(
                                                        ListFragment {
                                                            values: [
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "echo ",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "39.45",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " '+' ",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "21.18",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " | bc ",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "-l",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " | sed '",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "'",
                                                                    },
                                                                ),
                                                            ],
                                                            separator: Empty,
                                                        },
                                                    ),
                                                    quoted: true,
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " '+' ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "80.12",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | bc ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "-l",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | sed '",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "'",
                                                },
                                            ),
                                        ],
                                        separator: Empty,
                                    },
                                ),
                                quoted: true,
                            },
                        ),
                    ],
                    separator: Empty,
                },
            ),
            List(
                ListFragment {
                    values: [
                        Raw(
                            RawFragment {
                                value: "echo ",
                            },
                        ),
                        Subprocess(
                            SubprocessFragment {
                                fragment: List(
                                    ListFragment {
                                        values: [
                                            Raw(
                                                RawFragment {
                                                    value: "echo ",
                                                },
                                            ),
                                            Subprocess(
                                                SubprocessFragment {
                                                    fragment: List(
                                                        ListFragment {
                                                            values: [
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "echo ",
                                                                    },
                                                                ),
                                                                Subprocess(
                                                                    SubprocessFragment {
                                                                        fragment: List(
                                                                            ListFragment {
                                                                                values: [
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "echo ",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "3.0",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: " '+' ",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "9.0",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: " | bc ",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "-l",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: " | sed '",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                                                        },
                                                                                    ),
                                                                                    Raw(
                                                                                        RawFragment {
                                                                                            value: "'",
                                                                                        },
                                                                                    ),
                                                                                ],
                                                                                separator: Empty,
                                                                            },
                                                                        ),
                                                                        quoted: true,
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " '+' ",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "648.0",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " | bc ",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "-l",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: " | sed '",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                                    },
                                                                ),
                                                                Raw(
                                                                    RawFragment {
                                                                        value: "'",
                                                                    },
                                                                ),
                                                            ],
                                                            separator: Empty,
                                                        },
                                                    ),
                                                    quoted: true,
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " '+' ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "12.23",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | bc ",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "-l",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: " | sed '",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "/\\./ s/\\.\\{0,1\\}0\\{1,\\}$//",
                                                },
                                            ),
                                            Raw(
                                                RawFragment {
                                                    value: "'",
                                                },
                                            ),
                                        ],
                                        separator: Empty,
                                    },
                                ),
                                quoted: true,
                            },
                        ),
                    ],
                    separator: Empty,
                },
            ),
        ],
        increase_indent: true,
        needs_noop: false,
        is_conditional: false,
    },
)
